waftools/detections/compiler_swift.py

```diff
@@ -19,50 +21,162 @@ def __add_swift_flags(ctx):
+    ])
+
 def __find_swift_library(ctx):
-    swift_library_paths = [
-        'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
-        'usr/lib/swift_static/macosx'
-    ]
+    swift_libraries = {}
+    #look for set lib paths in passed environment variables
+    if 'SWIFT_LIB_DYNAMIC' in ctx.environ:
+        swift_libraries['SWIFT_LIB_DYNAMIC'] = ctx.environ['SWIFT_LIB_DYNAMIC']
+    if 'SWIFT_LIB_STATIC' in ctx.environ:
+        swift_libraries['SWIFT_LIB_STATIC'] = ctx.environ['SWIFT_LIB_STATIC']
+
+    #search for swift libs relative to the swift compiler executable
+    swift_library_relative_paths = {
+        'SWIFT_LIB_DYNAMIC': '../../lib/swift/macosx',
+        'SWIFT_LIB_STATIC': '../../lib/swift_static/macosx'
+    }
+
+    for lib_type, path in swift_library_relative_paths.items():
+        if lib_type not in swift_libraries:
+            lib_path = os.path.join(ctx.env.SWIFT, path)
+            swift_library = ctx.root.find_dir(lib_path)
+            if swift_library is not None:
+                swift_libraries[lib_type] = swift_library.abspath()
+
+    #fall back to xcode-select path
+    swift_library_paths = {
+        'SWIFT_LIB_DYNAMIC': [
+            'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx',
+            'usr/lib/swift/macosx'
+        ],
+        'SWIFT_LIB_STATIC': [
+            'Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx',
+            'usr/lib/swift_static/macosx'
+        ]
+    }
     dev_path = __run(['xcode-select', '-p'])[1:]
 
-    ctx.start_msg('Checking for Swift Library')
-    for path in swift_library_paths:
-        swift_library = ctx.root.find_dir([dev_path, path])
-        if swift_library is not None:
-            ctx.end_msg(swift_library.abspath())
-            __add_swift_library_linking_flags(ctx, swift_library.abspath())
-            return
-    ctx.end_msg(False)
+    for lib_type, paths in swift_library_paths.items():
+        for path in paths:
+            if lib_type not in swift_libraries:
+                swift_library = ctx.root.find_dir([dev_path, path])
+                if swift_library is not None:
+                    swift_libraries[lib_type] = swift_library.abspath()
+                    break
+            else:
+                break
+
+    #check if library paths were found
+    ctx.start_msg('Checking for dynamic Swift Library')
+    if 'SWIFT_LIB_DYNAMIC' in swift_libraries:
+        ctx.end_msg(swift_libraries['SWIFT_LIB_DYNAMIC'])
+    else:
+        ctx.end_msg(False)
+
+    ctx.start_msg('Checking for static Swift Library')
+    if 'SWIFT_LIB_STATIC' in swift_libraries:
+        ctx.end_msg(swift_libraries['SWIFT_LIB_STATIC'])
+        ctx.env['SWIFT_LIB_STATIC'] = swift_libraries['SWIFT_LIB_STATIC']
+    else:
+        ctx.end_msg(False)
+
+    enableStatic = getattr(ctx.options, 'enable_swift-static')
+    if (enableStatic) and 'SWIFT_LIB_STATIC' in swift_libraries:
```

Comment:
This kind of confuzzled me for a bit since `SWIFT_LIB_STATIC` was supposedly already checked with the `swift-static` feature. But apparently this stuff is run before the `func` defined, so we can end up here without the environment variable being defined. So in that sense some sort of checking makes sense, so that you don't end up with KeyError before properly failing to not have that environment variable defined.

That said, now we get the cases that:
* In case of `swift-static` being enabled, even if this part succeeds (static libs found through the search f.ex.) it still requires the environment variable in the `func` check.
* In case of `swift-static` being enabled, this code ends up silently adding the dynamic linking flags in case the environment variable is not set and the directories are not found. Of course, it will then later fail in the `func` so this I think is less of a problem. Just a weird case of success here compared to the other part.